new file:   News Scrapper UI.py

Aliminium-11929 opened the pull request:
created a tkinter app that pulls the 5 most recent news from almanar.com, then passes the content of the news to gemini llm api which summarizes the news and returns it to me, which I display on the app
![image](https://github.com/user-attachments/assets/33b139ed-321d-4f94-beab-50d3d5d27573)

HadiKotaich commented on News Scrapper UI.py:
```diff
@@ -0,0 +1,120 @@
+import arabic_reshaper
+from bidi.algorithm import get_display
+import textwrap
+
```
I would split the UI and the backend into two separate files if possible.

HadiKotaich commented on News Scrapper UI.py:
```diff
@@ -0,0 +1,120 @@
+google_gemini_api_key = os.getenv("GOOGLE_GEMINI_API_KEY")
+if not google_gemini_api_key:
+    raise ValueError("Google Gemini API key not set. Check your .env file.")
+feed_url = os.getenv("FEED_URL")
```
no need to make the FeedURL an environment variable, you can take it from the user.
or you can hardcode it here in the code, there is nothing seceret about it like the API key for example.

HadiKotaich commented on News Scrapper UI.py:
```diff
@@ -0,0 +1,120 @@
+# Function to fetch and display feed
+def get_feed_ui():
+    output_text.delete("1.0", END)  # Clear previous output
+    for entry in feed.entries[:5]:
```
you can try doing these in parallel, this is a great opportunity to learn async.

HadiKotaich reviewed (commented):
Looks good overall!! Thanks for working on this!

Aliminium-11929 commented:
I think I've fixed the commented areas on this branch now

HadiKotaich commented:
> I think I've fixed the commented areas on this branch now

Awesome, usually when comments are addressed, you can click on the resolve button, or reply on it if you have a question.

HadiKotaich commented on News_Scrapper_Backend.py:
```diff
@@ -40,37 +41,36 @@ def process_arabic_text(text, width=70):
+
+        if not article_content:
+            output_text=(f"Title: {title}\nLink: {link}\nPublished: {published}\n{'-' * 80}\n")
+            Arr[i]=output_text
```
be careful in multi-threading when using a global variable (or a common variable).

Usually when multiple threads are touching the same memory, it is advisable to use a lock.
This lock will serve as a way to "lock the resource" (i.e. not allow anyone else to use the global variable) if the lock is being used. So each thread will try to acquire the lock, if it succeeds, all other threads which are trying to acquire the lock will have to wait until the lock is released. 

In your case there was no issue because every thread is using a separate entry in the array, however this is usually tricky to manage.

HadiKotaich commented on News_Scrapper_Backend.py:
```diff
@@ -12,21 +12,22 @@
+feed_url = "https://almanar.com.lb/rss"
 feed = feedparser.parse(feed_url)
 
+Arr=["","","","",""]
```
I would avoid defining global arrays.
You can always create the array in the function and pass it as an argument to the function.

HadiKotaich commented on News_Scrapper_Backend.py:
```diff
@@ -40,37 +41,36 @@ def process_arabic_text(text, width=70):
+            output_text=(f"Title: {title}\nLink: {link}\nPublished: {published}\n{'-' * 80}\n")
+            Arr[i]=output_text
+
+        text = article_content.get_text(strip=True, separator="\n")
```
we would still get here if not article_content right? you are not returning above.

HadiKotaich commented on News_Scrapper_Backend.py:
```diff
@@ -40,37 +41,36 @@ def process_arabic_text(text, width=70):
+        text = article_content.get_text(strip=True, separator="\n")
+        if text:
+            reshaped_text = process_arabic_text(text, width=70)
+            content = google_gemini_query(reshaped_text)
```
how is content different from text.
we should use clear variable names and avoid generic ones.
in this case I think you mean "summary".

HadiKotaich commented on News_Scrapper_Backend.py:
```diff
@@ -12,21 +12,22 @@
 
-# Load environment variables
+
 load_dotenv()
```
I think those should be part of the google function

HadiKotaich commented on News_Scrapper_Backend.py:
```diff
@@ -40,37 +41,36 @@ def process_arabic_text(text, width=70):
+        Arr[i]=output_text
 
 # Google Gemini query function
 def google_gemini_query(prompt):
```
I don't think this is a good name for this function.
A better name would be: "summarize_text" or "summarize_text_gemeni"

HadiKotaich commented on News_Scrapper_UI.py:
```diff
@@ -0,0 +1,46 @@
+import os
```
Good spllit :)

HadiKotaich reviewed (commented):
Much better, good job doing the multi-threading way.
It is good to take a small look at what race condition mean and how to deal with them using locks. No need to go very deep.

Aliminium-11929 commented:
I added a new file version of the backend that addresses the above issues while keeping the old version for reference. The new version addresses the issues while also writing the data to a text file and can thus be called without a python UI. I'm hoping this should make it easier to work with a javascript UI.

Aliminium-11929 commented:
Hoping that I might be able to fix the array definition for the project sometime soon by deciding on a good number of articles to pull from each site.
